File: 0187. Repeated DNA Sequences.java
// Medium
// Sliding Window, Hash Set
// Time:O(n),Space:O(n)
// https://leetcode.cn/problems/repeated-dna-sequences/

import java.util.*;
class Solution {
    public List<String> findRepeatedDnaSequences(String s) {
        Set<String> res = new HashSet<>();
        Set<String> occur = new HashSet<>();
        for (int i = 0; i + 10 <= s.length(); i++) {
            String sub = s.substring(i, i + 10);
            // if substring occured before, add it to res
            if (occur.contains(sub)) {
                res.add(sub);
            } else {
                // or, add it to occur set
                occur.add(sub);
            }
        }
        return new ArrayList<>(res);
    }
}

File: 0001-0500/0187. Repeated DNA Sequences.py
# Medium
# String
# Time:O(n), Space:O(n)
# https://leetcode.cn/problems/repeated-dna-sequences/

from typing import List

class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        if len(s) < 10:
            return []
        
        seen = set()
        res = set()

        for i in range(len(s) - 9):
            sub_str = s[i:i+10]
            if sub_str in seen:
                res.add(sub_str)
            else:
                seen.add(sub_str)

        return list(res)
    